### ISSUE TITLE
Rule change request: no-async-promise-executor should be if-async-promise-executor-then-try-catch-surrounding-await

### ISSUE THREAD

**Issue by lropero** [eslint/eslint#11982]
**What rule do you want to change?** [no-async-promise-executor](https://eslint.org/docs/rules/no-async-promise-executor)

**Does this change cause the rule to produce more or fewer warnings?** Same result with fewer warnings

**How will the change be implemented? (New option, new default behavior, etc.)?** New default behavior

**Please provide some example code that this change will affect:**

```js
new Promise(async (resolve, reject) => {
  try {
    const someValue = await someAsyncMethod()
    const someOtherValue = await someOtherAsyncMethod()
    return resolve(someValue + someOtherValue)
  } catch (error) {
    return reject(error)
  }
})
```

**What does the rule currently do for this code?** Warns with `Promise executor functions should not be async.`

**What will the rule do after it's changed?** I think this rule should allow async executors whenever the `await` calls are within a try/catch block and promise is resolved/rejected accordingly, something like `if-async-promise-executor-then-try-catch-surrounding-await`. As it is, this rule is forcing me to write the above code using `.then()` constructs which I don't prefer. This will still prevent situations where async executors falls into the wrong pattern (i.e. resolving/rejecting synchronously).

**Are you willing to submit a pull request to implement this change?** Yes

**Comment by josezulu** [comment]
I agree with the original post: if the "await" is enclosed in a `try block` and the `catch` `rejects` the promise, errors thrown would not be missed. As far as I understood it this is regarded as good practice.

**Comment by not-an-aardvark** [comment]
Could you replace your code with this instead?

```js
(async (resolve, reject) => {
  const someValue = await someAsyncMethod()
  const someOtherValue = await someOtherAsyncMethod()
  return someValue + someOtherValue
})()
```

That would have the same behavior. It seems like using the `Promise` constructor is unnecessary if you're already using async functions to produce the result.

**Comment by lropero** [comment]
@not-an-aardvark How would I go about if I need to await this piece of code and bubble up any errors? For example, someOtherAsyncMethod() fails.

**Comment by not-an-aardvark** [comment]
If an error is thrown in an async function, the Promise that it returns will be rejected. So you will end up with the same functionality as your original code sample.

**Comment by lropero** [comment]
@not-an-aardvark I'm sorry but I still don't understand. 

In this example code I've written, how would I rewrite `mainAsyncMethod1()` and `mainAsyncMethod2()` to have the same implementation I have in `run()` method?

```
const mainAsyncMethod1 = () => new Promise(async (resolve, reject) => {
  try {
    const someValue = await someAsyncMethod()
    const someOtherValue = await someOtherAsyncMethod()
    return resolve(someValue + someOtherValue)
  } catch (error) {
    return reject(error)
  }
})

const mainAsyncMethod2 = () => new Promise(async (resolve, reject) => {
  try {
    const someValue = await someAsyncMethod()
    const someOtherValue = await someOtherAsyncMethod()
    return resolve(someValue + someOtherValue)
  } catch (error) {
    return reject(error)
  }
})

const run = async () => {
  try {
    const value1 = await mainAsyncMethod1()
    const value2 = await mainAsyncMethod2()
    console.log(\`Result is ${value1 + value2}\`)
  } catch (error) {
    console.log(error)
    process.exit(1)
  }
}

run()
```

**Comment by not-an-aardvark** [comment]
```js
const mainAsyncMethod1 = async () => {
  const someValue = await someAsyncMethod()
  const someOtherValue = await someOtherAsyncMethod()
  return someValue + someOtherValue;
}

const mainAsyncMethod2 = async () => {
  const someValue = await someAsyncMethod()
  const someOtherValue = await someOtherAsyncMethod()
  return someValue + someOtherValue
}

const run = async () => {
  try {
    const value1 = await mainAsyncMethod1()
    const value2 = await mainAsyncMethod2()
    console.log(`Result is ${value1 + value2}`)
  } catch (error) {
    console.log(error)
    process.exit(1)
  }
}

run()
```

**Comment by lropero** [comment]
I guess you're right, just tried this and returns **10** after 4 seconds as expected:

```js
const mainAsyncMethod1 = async () => {
  const someValue = await new Promise((resolve, reject) => setTimeout(() => resolve(1), 1000))
  const someOtherValue = await new Promise((resolve, reject) => setTimeout(() => resolve(2), 1000))
  return someValue + someOtherValue
}

const mainAsyncMethod2 = async () => {
  const someValue = await new Promise((resolve, reject) => setTimeout(() => resolve(3), 1000))
  const someOtherValue = await new Promise((resolve, reject) => setTimeout(() => resolve(4), 1000))
  return someValue + someOtherValue
}

const run = async () => {
  try {
    const value1 = await mainAsyncMethod1()
    const value2 = await mainAsyncMethod2()
    console.log(`Result is ${value1 + value2}`)
  } catch (error) {
    console.log(error)
    process.exit(1)
  }
}

run()
```

Thank you for clarifying.
